src/ossl/ecdh.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright 2023 - 2024 Simo Sorce, Jakub Jelen
+// See LICENSE.txt file for terms
+
+use core::ffi::{c_char, c_int, c_uint};
+use std::borrow::Cow;
+
+use crate::attribute::CkAttrs;
+use crate::bytes_to_vec;
+use crate::ecc::*;
+use crate::error::Result;
+use crate::interface::*;
+use crate::mechanism::*;
+use crate::object::{default_key_attributes, Object, ObjectFactories};
+use crate::ossl::bindings::*;
+use crate::ossl::common::*;
+#[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+use crate::ossl::ec_montgomery::{
+    is_montgomery_curve, make_ec_montgomery_public_key,
+    make_output_length_from_montgomery_obj,
+    montgomery_object_to_ecc_private_key,
+};
+#[cfg(feature = "ecc")]
+use crate::ossl::ecc::{
+    ecdsa_object_to_ecc_private_key, get_curve_name_from_obj,
+    make_ecdsa_public_key, make_output_length_from_ecdsa_obj,
+};
+
+fn kdf_type_to_hash_mech(mech: CK_EC_KDF_TYPE) -> Result<CK_MECHANISM_TYPE> {
+    match mech {
+        CKD_SHA1_KDF => Ok(CKM_SHA_1),
+        CKD_SHA224_KDF => Ok(CKM_SHA224),
+        CKD_SHA256_KDF => Ok(CKM_SHA256),
+        CKD_SHA384_KDF => Ok(CKM_SHA384),
+        CKD_SHA512_KDF => Ok(CKM_SHA512),
+        CKD_SHA3_224_KDF => Ok(CKM_SHA3_224),
+        CKD_SHA3_256_KDF => Ok(CKM_SHA3_256),
+        CKD_SHA3_384_KDF => Ok(CKM_SHA3_384),
+        CKD_SHA3_512_KDF => Ok(CKM_SHA3_512),
+        _ => return Err(CKR_MECHANISM_PARAM_INVALID)?,
+    }
+}
+
+fn make_derive_output_length_from_obj(key: &Object) -> Result<usize> {
+    // Montgomery
+    #[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+    if is_montgomery_curve(key) {
+        return make_output_length_from_montgomery_obj(key);
+    }
+    // ECDSA
+    make_output_length_from_ecdsa_obj(key)
+}
+
+fn derive_object_to_ecc_private_key(key: &Object) -> Result<EvpPkey> {
+    // Montgomery
+    #[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+    if is_montgomery_curve(key) {
+        return montgomery_object_to_ecc_private_key(key);
+    }
+    // ECDSA
+    ecdsa_object_to_ecc_private_key(key)
+}
+
+fn make_peer_key(key: &Object, ec_point: &Vec<u8>) -> Result<EvpPkey> {
+    // Montgomery
+    #[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+    if is_montgomery_curve(key) {
+        return make_ec_montgomery_public_key(key, ec_point);
+    }
+    // ECDSA
+    match get_curve_name_from_obj(key) {
+        Ok(c) => make_ecdsa_public_key(&c, &ec_point),
+        Err(_) => return Err(CKR_DEVICE_ERROR)?,
+    }
+}
+
+#[derive(Debug)]
+pub struct ECDHOperation {
+    mech: CK_MECHANISM_TYPE,
+    kdf: CK_EC_KDF_TYPE,
+    public: Vec<u8>,
+    shared: Vec<u8>,
+    finalized: bool,
+}
+
+impl ECDHOperation {
+    fn new_mechanism() -> Box<dyn Mechanism> {
```

Comment:
I think that new_mechanism() and register_mechanisms() should be moved into a src/ecdh.rs file. This is just like HKDFMechanism and should mimic that mechanism derive construction.

Reply:
Thank you for the suggestion. It clearly makes sense. Done.